mongocrypt/build.gradle.kts

```diff
@@ -0,0 +1,197 @@
+val jnaResourcesDir = "$buildDir/jnaLibs/resources/"
+val jnaLibPlatform: String = if (com.sun.jna.Platform.RESOURCE_PREFIX.startsWith("darwin")) "darwin" else com.sun.jna.Platform.RESOURCE_PREFIX
+val jnaLibsPath: String = System.getProperty("jnaLibsPath", "${jnaResourcesDir}${jnaLibPlatform}")
+val jnaResources: String = System.getProperty("jna.library.path", jnaLibsPath)
+
+// Download jnaLibs that match the git tag or revision to jnaResourcesBuildDir
+val downloadRevision = "9a88ac5698e8e3ffcd6580b98c247f0126f26c40" // r.1.11.0
+val binariesArchiveName = "libmongocrypt-java.tar.gz"
+
+/**
+ * The name of the archive includes downloadRevision to ensure that:
+ * - the archive is downloaded if the revision changes.
+ * - the archive is not downloaded if the revision is the same and archive had already been saved in build output.
+ */
+val localBinariesArchiveName = "libmongocrypt-java-$downloadRevision.tar.gz"
+
+val downloadUrl: String = "https://mciuploads.s3.amazonaws.com/libmongocrypt/java/$downloadRevision/$binariesArchiveName"
+
+val jnaMapping: Map<String, String> = mapOf(
+    "rhel-62-64-bit" to "linux-x86-64",
+    "rhel72-zseries-test" to "linux-s390x",
+    "rhel-71-ppc64el" to "linux-ppc64le",
+    "ubuntu1604-arm64" to "linux-aarch64",
+    "windows-test" to "win32-x86-64",
+    "macos" to "darwin"
+)
+
+sourceSets {
+    main {
+        java {
+            resources {
+                srcDirs(jnaResourcesDir)
+            }
+        }
+    }
+}
+
+tasks.register<Download>("downloadJava") {
+    src(downloadUrl)
+    dest("${jnaDownloadsDir}/$localBinariesArchiveName")
+    overwrite(true)
+    /* To make sure we don't download archive with binaries if it hasn't been changed in S3 bucket since last download.*/
+    onlyIfModified(true)
+}
+
+tasks.register<Copy>("unzipJava") {
+    /*
+        Clean up the directory first if the task is not UP-TO-DATE.
+        This can happen if the download revision has been changed and the archive is downloaded again.
+     */
+    doFirst {
+        println("Cleaning up $jnaResourcesDir")
+        delete(fileTree(jnaResourcesDir).matching {
+            include(jnaMapping.keys.flatMap {
+                listOf(
+                    "${it}/nocrypto/**/libmongocrypt.so",
+                    "${it}/lib/**/libmongocrypt.dylib",
+                    "${it}/bin/**/mongocrypt.dll"
+                )
+            })
+        })
+    }
+    from(tarTree(resources.gzip("${jnaDownloadsDir}/$localBinariesArchiveName")))
+    include(jnaMapping.keys.flatMap {
+        listOf("${it}/nocrypto/**/libmongocrypt.so", "${it}/lib/**/libmongocrypt.dylib", "${it}/bin/**/mongocrypt.dll" )
+    })
+    eachFile {
+        path = "${jnaMapping[path.substringBefore("/")]}/${name}"
+    }
+    into(jnaResourcesDir)
+    mustRunAfter("downloadJava")
+
+    doLast {
+        println("jna.library.path contents: \n  ${fileTree(jnaResourcesDir).files.joinToString(",\n  ")}")
+    }
+}
```

Comment:
Gradle tests whether task `inputs` or `outputs` have changed since the last build as part of the incremental build process. If no changes are detected, it considers the task up-to-date and skips execution.

There are two time-consuming tasks now using incremental builds to improve build efficiency:

- downloadJava: Downloads `libmongocrypt` binaries from S3. Enhanced with the `onlyIfModified(true)` option, this task avoids redundant downloads by fetching binaries only if they are absent from the build output.

The `localBinariesArchiveName` variable in the `./build` folder, containing the zip archive file name with the revision number, ensures that binaries are downloaded just once per revision/tag unless a `gradle clean` command explicitly clears the build folder.

- unzipJava: Unzips binaries into the temporary resource folder. Previously, this task was configured with `outputs.upToDateWhen { false }` to always execute. Now, it triggers only under these conditions:
    - The input file name changes, detected from the `downloadRevision` variable, prompting the unzipping of the latest or previously saved binaries.
    - Any alteration or corruption in the output file structure within `jnaResourcesDir`. To ensure correct unzipping, it first cleans `jnaResourcesDir` and then proceeds to unzip the binaries.

With `mongocrypt` as a `project dependency` of `driver-core`, the incremental build process conserves time when running or testing code unrelated to the `mongocrypt module` in sync/async modules. It avoids unnecessary rebuilding of the `mongocrypt jar` for the `driver-core module`. This is also useful during active development as `libmongocrypt` commits may update frequently.